src/main/java/org/tikv/common/util/ChannelFactory.java

```diff
@@ -215,63 +216,69 @@ public ChannelFactory(
     this.certContext = new JksContext(jksKeyPath, jksKeyPassword, jksTrustPath, jksTrustPassword);
   }
 
-  @VisibleForTesting
-  public boolean reloadSslContext() {
-    if (certContext != null) {
-      SslContextBuilder newBuilder = certContext.reload();
-      if (newBuilder != null) {
-        sslContextBuilder.set(newBuilder);
-        return true;
+  private ManagedChannel createChannel(
+      SslContextBuilder sslContextBuilder, String address, HostMapping mapping) {
+    URI uri, mapped;
+    try {
+      uri = PDUtils.addrToUri(address);
+    } catch (Exception e) {
+      throw new IllegalArgumentException("failed to form address " + address, e);
+    }
+    try {
+      mapped = mapping.getMappedURI(uri);
+    } catch (Exception e) {
+      throw new IllegalArgumentException("failed to get mapped address " + uri, e);
+    }
+
+    // Channel should be lazy without actual connection until first call
+    // So a coarse grain lock is ok here
+    NettyChannelBuilder builder =
+        NettyChannelBuilder.forAddress(mapped.getHost(), mapped.getPort())
+            .maxInboundMessageSize(maxFrameSize)
+            .keepAliveTime(keepaliveTime, TimeUnit.SECONDS)
+            .keepAliveTimeout(keepaliveTimeout, TimeUnit.SECONDS)
+            .keepAliveWithoutCalls(true)
+            .idleTimeout(idleTimeout, TimeUnit.SECONDS);
+
+    if (sslContextBuilder == null) {
+      return builder.usePlaintext().build();
+    } else {
+      SslContext sslContext;
+      try {
+        sslContext = sslContextBuilder.build();
+      } catch (SSLException e) {
+        logger.error("create ssl context failed!", e);
+        return null;
       }
+      return builder.sslContext(sslContext).build();
     }
-    return false;
   }
 
-  public ManagedChannel getChannel(String addressStr, HostMapping hostMapping) {
-    if (reloadSslContext()) {
-      logger.info("invalidate connection pool");
+  @VisibleForTesting
+  public synchronized ManagedChannel reload(String address, HostMapping mapping) {
```

Comment:
I am worried about the performance of getChannel(). Since renewing cert is a very low frequent operation, it's a big cost to sync and invoke a disk IO every time

Reply:
And from my point of view, client-java needs to have a high throughput when working as a cache client.
It's a situation quite like service discovery in RPC client, in which we wouldn't refresh every time getChannel

Reply:
> 

There is an optimization to avoid sync disk IO, using a `WatchService` to notify the `ChannelFactory` if the cert is changed. A TODO  is added.

Reply:
Would you enhance it recently?